geode-wan/src/main/java/org/apache/geode/cache/client/internal/locator/wan/LocatorMembershipListenerImpl.java

```diff
@@ -213,4 +199,119 @@ public void clearLocatorInfo() {
     allLocatorsInfo.clear();
     allServerLocatorsInfo.clear();
   }
+
+  private static class DistributeLocatorsRunnable implements Runnable {
+    private final int memberTimeout;
+    private final TcpClient tcpClient;
+    private final DistributionLocatorId localLocatorId;
+    private final Map<Integer, Set<DistributionLocatorId>> remoteLocators;
+    private final DistributionLocatorId joiningLocator;
+    private final int joiningLocatorDistributedSystemId;
+
+    DistributeLocatorsRunnable(int memberTimeout,
+        TcpClient tcpClient,
+        DistributionLocatorId localLocatorId,
+        Map<Integer, Set<DistributionLocatorId>> remoteLocators,
+        DistributionLocatorId joiningLocator,
+        int joiningLocatorDistributedSystemId) {
+
+      this.memberTimeout = memberTimeout;
+      this.tcpClient = tcpClient;
+      this.localLocatorId = localLocatorId;
+      this.remoteLocators = remoteLocators;
+      this.joiningLocator = joiningLocator;
+      this.joiningLocatorDistributedSystemId = joiningLocatorDistributedSystemId;
+    }
+
+    void sendMessage(DistributionLocatorId targetLocator, LocatorJoinMessage locatorJoinMessage,
+        Map<DistributionLocatorId, Set<LocatorJoinMessage>> failedMessages) {
+      DistributionLocatorId advertisedLocator = locatorJoinMessage.getLocator();
+
+      try {
+        tcpClient.requestToServer(targetLocator.getHost(), locatorJoinMessage, memberTimeout,
+            false);
+      } catch (Exception exception) {
+        if (logger.isDebugEnabled()) {
+          logger.debug(LISTENER_FAILURE_MESSAGE,
+              new Object[] {advertisedLocator.getHostName(), advertisedLocator.getPort(),
+                  targetLocator.getHostName(), targetLocator.getPort(), 1, exception});
+        }
+
+        if (!failedMessages.containsKey(targetLocator)) {
+          failedMessages.put(targetLocator, new HashSet<>());
+        }
+
+        failedMessages.get(targetLocator).add(locatorJoinMessage);
+      }
+    }
+
+    boolean retryMessage(DistributionLocatorId targetLocator, LocatorJoinMessage locatorJoinMessage,
+        int retryAttempt) {
+      DistributionLocatorId advertisedLocator = locatorJoinMessage.getLocator();
+
+      try {
+        tcpClient.requestToServer(targetLocator.getHost(), locatorJoinMessage, memberTimeout,
+            false);
+
+        return true;
+      } catch (Exception exception) {
+        if (retryAttempt == WAN_LOCATOR_DISTRIBUTION_RETRY_ATTEMPTS) {
+          logger.warn(LISTENER_FINAL_FAILURE_MESSAGE,
```

Comment:
Do we need to warn during retry or warn after the retry is completed and there is a failure. Sometimes customer scrub logs for warn messages.

Reply:
We need to let the user know that the exchange failed from this locator to the remote locator, that's why only when all the retry attempts fail the `warn` level is used, the rest of the messages are only logged using `debug` level.